plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/CancellableRateLimitedFluxIterator.java

```diff
@@ -0,0 +1,230 @@
+ * invoked for all the elements that weren't consumed before the cancellation. (i.e. it's
+ * possible to free the memory allocated for a byte buffer).
+ */
+class CancellableRateLimitedFluxIterator<T> implements Subscriber<T>, Iterator<T> {
+    private static final Subscription CANCELLED_SUBSCRIPTION = new Subscription() {
+        @Override
+        public void request(long n) {
+            // no op
+        }
+
+        @Override
+        public void cancel() {
+            // no op
+        }
+    };
+
+    private final int elementsPerBatch;
+    private final Queue<T> queue;
+    private final Lock lock;
+    private final Condition condition;
+    private final Consumer<T> cleaner;
+    private final AtomicReference<Subscription> subscription = new AtomicReference<>();
+    private final Logger logger = LogManager.getLogger(CancellableRateLimitedFluxIterator.class);
+    private volatile Throwable error;
+    private volatile boolean done;
+    private int emittedElements;
+
+    /**
+     * Creates a new CancellableRateLimitedFluxIterator that would request to it's upstream publisher
+     * in batches as specified in {@code elementsPerBatch}. Additionally, it's possible to provide a
+     * function that would be invoked after cancellation for possibly outstanding elements that won't by
+     * consumed downstream but need to be cleaned in any case.
+     * @param elementsPerBatch the number of elements to request upstream
+     * @param cleaner the function that would be used to clean unused elements
+     */
+    CancellableRateLimitedFluxIterator(int elementsPerBatch, Consumer<T> cleaner) {
+        this.elementsPerBatch = elementsPerBatch;
+        this.queue = new ArrayDeque<>(elementsPerBatch);
+        this.lock = new ReentrantLock();
+        this.condition = lock.newCondition();
+        this.cleaner = cleaner;
+    }
+
+    @Override
+    public boolean hasNext() {
+        for (; ; ) {
+            boolean isDone = done;
+            boolean isQueueEmpty = queue.isEmpty();
+
+            if (isDone) {
+                Throwable e = error;
+                if (e != null) {
+                    throw new RuntimeException(e);
+                } else if (isQueueEmpty) {
+                    return false;
+                }
+            }
+
+            if (isQueueEmpty == false) {
+                return true;
+            }
+
+            // Provide visibility guarantees for the modified queue
+            lock.lock();
+            try {
+                while (done == false && queue.isEmpty()) {
+                    condition.await();
+                }
+            } catch (InterruptedException e) {
+                cancelSubscription();
+                throw new RuntimeException(e);
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    @Override
+    public T next() {
+        if (hasNext() == false) {
```

Comment:
Can we document that `hasNext()` is where we block and wait for the IO thread to put something in the queue? This is quite subtle I think :)

Reply:
Fixed in https://github.com/elastic/elasticsearch/pull/65140/commits/2c3eefa1117b3371a21f42ea7902058f274fb746